contracts/handlers/NativeTokenHandler.sol

```diff
@@ -65,21 +79,68 @@ contract NativeTokenHandler is IHandler, ERCHandlerHelpers {
+        optionalMessage             length uint256 bytes (96 + len(destinationRecipientAddress)) - (128 + len(destinationRecipientAddress))
+        optionalMessage                    bytes   bytes (160 + len(destinationRecipientAddress)) - END
      */
     function executeProposal(bytes32 resourceID, bytes calldata data) external override onlyBridge returns (bytes memory) {
-        (uint256 amount) = abi.decode(data, (uint256));
+        uint256 amount;
+        uint256 lenDestinationRecipientAddress;
+
+        (amount, lenDestinationRecipientAddress) = abi.decode(data, (uint256, uint256));
+        address recipientAddress = address(bytes20(bytes(data[64:64 + lenDestinationRecipientAddress])));
+
         address tokenAddress = _resourceIDToTokenContractAddress[resourceID];
-        address recipientAddress = address(bytes20(bytes(data[64:84])));
-        uint256 convertedAmount = convertToExternalBalance(tokenAddress, amount);
 
-        (bool success, ) = address(recipientAddress).call{value: convertedAmount}("");
-        if(!success) revert FailedFundsTransfer();
-        emit FundsTransferred(recipientAddress, amount);
+        // Optional message recipient transformation.
+        uint256 pointer = 64 + lenDestinationRecipientAddress;
+        uint256 gas;
+        uint256 messageLength;
+        OptionalMessageCheck optionalMessageCheck;
+        if (data.length > (pointer + 64)) {
+            (gas, messageLength) = abi.decode(data[pointer:], (uint256, uint256));
+            pointer += 64;
+            if (gas > 0 && messageLength > 0 && (messageLength + pointer) <= data.length) {
+                optionalMessageCheck = OptionalMessageCheck.Valid;
+                if (recipientAddress == transformRecipient) {
+                    recipientAddress = _defaultMessageReceiver;
+                }
+            } else {
+                gas = defaultGas;
+                optionalMessageCheck = OptionalMessageCheck.Invalid;
+            }
+        }
+
+        if (optionalMessageCheck == OptionalMessageCheck.Invalid) {
+            return abi.encode(
+                tokenAddress,
+                recipientAddress,
+                amount,
+                abi.encode(false, abi.encodeWithSignature("InvalidEncoding()"))
+            );
+        }
+
+        uint256 externalAmount = convertToExternalBalance(tokenAddress, amount);
+        bytes memory recipientMessage = "";
+        if (optionalMessageCheck == OptionalMessageCheck.Valid) {
+            bytes memory message = bytes(data[pointer:pointer + messageLength]);
+            recipientMessage = abi.encodeWithSelector(
+                ISygmaMessageReceiver(recipientAddress).handleSygmaMessage.selector,
+                tokenAddress,
+                externalAmount,
+                message
+            );
+        }
+        (bool success, bytes memory result) =
+            recipientAddress.excessivelySafeCall(gas, externalAmount, maxReturnBytes, recipientMessage);
```

Comment:
For native token it's failing in case of value(token amount) is greater than zero. [here](https://dashboard.tenderly.co/tx/base-sepolia/0x5f76faa7c000141bd1895440628826c50909f43ea0a1795248491897d854d421?trace=0)
But if the value is zero(while depositing if I just set amount = sygma fee) then it got successful. [here](https://dashboard.tenderly.co/tx/base-sepolia/0x3d549f636085bc4bc481ff5cab93995b6c286746c3108e369e4b458893c258aa)

Reply:
Something is confused here. First link points to the execution of ERC20Handler proposal and the externalAmount is 0 there.
For NativeTokenHandler it could fail if the receiver is a contract which fail to receive native token. Could you add a link to the failed NativeTokenHandler execution with positive externalAmount?

Reply:
Seems like the issue was regarding lack of eth on native handler

Reply:
You can mark as resloved